### DISCUSSION TITLE
instance not found

### DISCUSSION THREAD

**Discussion by abrahamhwj** [kubernetes-sigs/kubebuilder#3171] [Support] [answered]
I just newer to learn this, I hop get some help, 3ks
---
when I run my controller, the log show this:

> 1.67401026673349e+09    INFO    no xcoreDefinition instance found.      {"controller": "xcoredefinition", "controllerGroup": "xcore.xg.com", "controllerKind": "XcoreDefinition", "xcoreDefinition": {"name":"xcoredefinition-sample"}, "namespace": "
> ", "name": "xcoredefinition-sample", "reconcileID": "8583b4b5-7883-4da7-b446-78d7e9eed081"}
> 1.6740102667348359e+09  INFO    no xcoreDefinition instance found.      {"controller": "xcoredefinition", "controllerGroup": "xcore.xg.com", "controllerKind": "XcoreDefinition", "xcoreDefinition": {"name":"xcore-sample"}, "namespace": "", "name":
>  "xcore-sample", "reconcileID": "d2945031-f38f-4f4d-a407-646cddd4f0f9"}
> 1.6740102667359676e+09  INFO    no xcoreDefinition instance found.      {"controller": "xcoredefinition", "controllerGroup": "xcore.xg.com", "controllerKind": "XcoreDefinition", "xcoreDefinition": {"name":"sample"}, "namespace": "", "name": "samp
> le", "reconcileID": "a52060e5-51c1-4c3c-bbac-c9076f7cc2e0"}

my question:
1、xcoredefinition-sample\xcore-sample, these 2 instances already deleted before controller run, where these info from?
2、"sample" instance  in  namespace "xg"，already created before controller run，but as you can see, the log noti "not found"; I need apply this instance again after controller runing, then the controller can find it, it is normal? or I miss something?

**Comment by camilamacedo86** [comment, marked as answer]
Hi @abrahamhwj,

Following some points that might help you out. 

1. **Regarding `xcoredefinition-sample` and `xcore-sample`:**
   - When an instance of a Custom Resource (CR) is deleted, the `controller-runtime`'s reconcile loop still triggers. This is often intended to ensure resources are cleaned up or to handle the deletion event. 
   - The events could be cached in your cluster's `etcd`, causing them to still be visible to the controller even after deletion. This cache can lead to "phantom" reconciliation events.
   
2. **Regarding the "sample" instance in the namespace "xg":**
   - It's possible that the controller started before the Custom Resource instance was recognized by the API server. Reapplying it forces the API server to recognize it, and then the controller can reconcile it.
   - It might also be related to how caching works in `controller-runtime`. The cache could be causing a delay in the controller's recognition of the instance.
   
### Further Information on Finalizers:
- For a deeper understanding of finalizers in Kubernetes, refer to the official [Kubernetes documentation on finalizers](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers).
- The `Deploy Image` plugin in `kubebuilder` provides an example of a finalizer. You can inspect its usage [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.12.0/testdata/project-v4-with-deploy-image/internal/controller/busybox_controller.go#L121-L192).
  
**Suggestion:** It might be helpful to review the content of this PR: [kubebuilder PR #3596](https://github.com/kubernetes-sigs/kubebuilder/pull/3596/files). It offers a comprehensive guide that might clarify how things operate.

I hope that answered your question
So, I am closing this one but feel free to reopen if you see that fits.
